### PR TITLE
Feature/blow down model

### PR DESCRIPTION
**Pull request by andrewmorell**
* **Tickets addressed:** MAXGNC-999
* **Review:** By commit
* **Merge strategy:** Merge (no squash)

## Description
Changes are made to fuelTank state effector, thrusterDynamicEffector, THROutputMsgPayload, and the THROperation message to enable toggling whether a fuel tank is subject to blow down effects, computing these blow down effects in the thrusterDynamicEffector if it is connected to such a tank, and implementing the BlowDownFactor in computing the thruster's force and torque.

## Verification
A parametrized unit test has been added to the existing thrusterDynamicEffector's unit test which checks the computeBlowDownDecay method, and the existing integrated test has also been updated to check correct setup of blow down coefficients when using the thruster factory and correct connection of a fuel tank's mass state to the thruster.

Verification of this new feature working with the ema-gnc scenarios has also been performed for the desat and closed loop DV scenarios. See plots in Slack showing blow down effects with the initial linear model.

## Documentation
The latex document for this module has been updated to include blow down effects, but also required adding some missing documentation on previously added swirl torque features.

## Future work
This BlowDownEffect could be extended to the thrusterStateEffector.

### CODE REVIEW

File: src/simulation/dynamics/FuelTank/fuelTank.h
Comment (codyallard28) [inline src/simulation/dynamics/FuelTank/fuelTank.h]: Since you made it explicit to be ThrusterDynamicEffectors and ThrusterStateEffectors, doesn't it make sense to change the variable names too?

File: src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.h
Comment (codyallard28) [inline src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.h]: This variable is used in "Add the fuel mass state to thrusterDynamicEffector" commit

File: src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp
Comment (codyallard28) [inline src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp]: can it->MaxThrust * ops->ThrustFactor * ops->BlowDownFactor be replaced by SingleThrusterForce?

File: src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp
Comment (codyallard28) [inline src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp]: I think this makes sense for the initial linear model. Zero thrust at zero mass percentage. We aren't giving an option to have a generic y = mx + b model, but I don't think it is necessary. Maybe just explicitly say what type of model this is in a comment?

**Review by codyallard28 — CHANGES_REQUESTED** [review]
Just a couple of minor changes before merging.

File: src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp
Comment (patkenneally) [inline src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp]: :small_red_triangle_down: The braces around the variable are not needed.

File: src/simulation/dynamics/_GeneralModuleFiles/THROperation.h
Comment (patkenneally) [inline src/simulation/dynamics/_GeneralModuleFiles/THROperation.h]: :small_red_triangle_down: `blowDownFactor`

File: src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp
Comment (patkenneally) [inline src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp]: :small_red_triangle_down: `computeBlowDownDecay()`

File: src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp
Comment (patkenneally) [inline src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp]: :small_red_triangle_down: `this->computeBlowDownDecay()`

File: src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp
Comment (patkenneally) [inline src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp]: :small_red_triangle_down: You can delete this line and allow the compiler to infer the type of the variable within the loop using `auto`.

**Review by patkenneally — CHANGES_REQUESTED** [review]
Legend:
:small_red_triangle_down: *Issues* to address before merge
:large_orange_diamond: *Requests* that should not block merge, but should at least be discussed
:large_blue_circle: *Recommendations* that can be ignored if desired

I would have expected some changes/additions to tests given the functionality changes. I'd be grateful if you could help me understand why there are no test changes?

File: src/simulation/dynamics/FuelTank/fuelTank.h
Comment (andrewmorell) [reply src/simulation/dynamics/FuelTank/fuelTank.h]: I agree, this has been changed now

File: src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.h
Comment (andrewmorell) [reply src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.h]: Ah yes, thanks for catching this. I've moved the fuel tank's use of the variable out of the "Add the fuel mass to thrusterDynamicEffector" commit and into the "Implement blow down in thrusterDynamicEffector" commit where it is created here on the thruster side.

File: src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp
Comment (andrewmorell) [reply src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp]: Good point, it can't be replaced by SingleThrusterForce but it can be replaced by tmpThrustMag for cleaner code so I've done that.

File: src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp
Comment (andrewmorell) [reply src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp]: Ultimately both this simplified version of the linear model and the new polynomial model given to us by structures could use the same code, so I'm implementing that.

**Comment by andrewmorell** [comment]
Sorry for the confusion with this being a PR not ready to be a PR yet, I'm converting it to a draft for now to hopefully avoid confusion while preserving what is already here.

File: src/simulation/dynamics/_GeneralModuleFiles/THROperation.h
Comment (andrewmorell) [reply src/simulation/dynamics/_GeneralModuleFiles/THROperation.h]: This has been changed to a list of thruster blow down polynomial coefficients

File: src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp
Comment (andrewmorell) [reply src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp]: Agreed

File: src/simulation/dynamics/FuelTank/fuelTank.h
Comment (patkenneally) [inline src/simulation/dynamics/FuelTank/fuelTank.h:266]: While not your doing, I wonder if these should be private. There's no direct usage of them throughout the project and we already have functions for adding an effector to the list, seems like we should just finish the job.

File: src/simulation/dynamics/FuelTank/fuelTank.cpp
Comment (patkenneally) [inline src/simulation/dynamics/FuelTank/fuelTank.cpp]: :large_orange_diamond: This can be passed in as a `const` reference e.g. `FuelTank::setNameOfMassState(const std::string& nameOfMassState)`

File: src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.h
Comment (patkenneally) [inline src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.h:62]: :large_blue_circle: Random new line appearing for no reason :)

File: src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp
Comment (patkenneally) [inline src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp]: :large_orange_diamond: We can move this to the initializer list already partially added

File: src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.h
Comment (patkenneally) [inline src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.h:71]: :large_orange_diamond: Should this be public? Let's discuss.

File: src/utilities/simIncludeThruster.py
Comment (patkenneally) [inline src/utilities/simIncludeThruster.py:175]: :small_red_triangle_down: We should be throwing an Exception here, not printing and exiting. Let's chat. I'm likely to delete these `*Includes.py` files soon anyhow.

File: src/simulation/dynamics/_GeneralModuleFiles/THROperation.h
Comment (patkenneally) [inline src/simulation/dynamics/_GeneralModuleFiles/THROperation.h]: :large_orange_diamond: Does this mean that it's defaulted to 1%?

File: src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp
Comment (patkenneally) [inline src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp]: :large_orange_diamond: What do you think of adding a line break to reduce this line length below 120 characters?

File: src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp
Comment (patkenneally) [inline src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp]: :large_orange_diamond: Long line

File: src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp
Comment (patkenneally) [inline src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp]: :large_orange_diamond: Long line

File: src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp
Comment (patkenneally) [inline src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp]: :large_orange_diamond: `CurrentThruster` to currentThruster`

File: src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp
Comment (patkenneally) [inline src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp]: 🔶 Long line

File: src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp
Comment (patkenneally) [inline src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp]: :small_red_triangle_down: Let's use the foreach syntax. E.g.

```
for(const auto& thrusterConfig : this->thrusterData) {
if (this->fuelMass < 0.0 && 
            (!thrusterConfig.thrBlowDownCoeff.empty() || !thrusterConfig.ispBlowDownCoeff.empty())) {
            bskLogger.bskLog(BSK_WARNING, "ThrusterDynamicEffector: blow down coefficients have been specified, "
                                          "but no fuel tank is attached.");
    }
}
```

File: src/simulation/dynamics/Thrusters/thrusterDynamicEffector/_UnitTest/test_thruster_integrated_sim.py
Comment (patkenneally) [inline src/simulation/dynamics/Thrusters/thrusterDynamicEffector/_UnitTest/test_thruster_integrated_sim.py]: :small_red_triangle_down: We consider `unitTestSupport` deprecated. Use numpy.testing asserts instead.

If possible we should convert this test numpy.testing asserts and delete all the `testFailCount` pieces.

File: src/simulation/dynamics/Thrusters/thrusterDynamicEffector/_UnitTest/test_ThrusterDynamicsUnit.py
Comment (patkenneally) [inline src/simulation/dynamics/Thrusters/thrusterDynamicEffector/_UnitTest/test_ThrusterDynamicsUnit.py:263]: I'm tempted to delete all this TeX from the test file. Let's chat.

**Review by patkenneally — APPROVED** [review]
Legend:
:small_red_triangle_down: *Issues* to address before merge
:large_orange_diamond: *Requests* that should not block merge, but should at least be discussed
:large_blue_circle: *Recommendations* that can be ignored if desired\

Solid PR. Complicated, but broken out nicely and so easy to follow.

File: src/simulation/dynamics/Thrusters/thrusterDynamicEffector/_UnitTest/test_ThrusterDynamicsUnit.py
Comment (patkenneally) [reply src/simulation/dynamics/Thrusters/thrusterDynamicEffector/_UnitTest/test_ThrusterDynamicsUnit.py:263]: I'll create a ticket.

File: src/utilities/simIncludeThruster.py
Comment (patkenneally) [reply src/utilities/simIncludeThruster.py:175]: I'll create a ticket for this.

File: src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.h
Comment (patkenneally) [reply src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.h:71]: I'll create a ticket for a complete module overhaul.

File: src/simulation/dynamics/FuelTank/fuelTank.h
Comment (patkenneally) [reply src/simulation/dynamics/FuelTank/fuelTank.h:266]: This will be part of module overhaul ticket.

File: src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.h
Comment (andrewmorell) [reply src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.h:62]: Understood now to move these standalone formatting changes into their own commit.

File: src/simulation/dynamics/FuelTank/fuelTank.cpp
Comment (andrewmorell) [reply src/simulation/dynamics/FuelTank/fuelTank.cpp]: This makes sense, done.

File: src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp
Comment (andrewmorell) [reply src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp]: Agreed, both mDotTotal and fuelMass are now moved to the initializer list.

File: src/simulation/dynamics/_GeneralModuleFiles/THROperation.h
Comment (andrewmorell) [reply src/simulation/dynamics/_GeneralModuleFiles/THROperation.h]: This means it is defaulting to 100%, comment now fixed to reflect that this is a scaling factor rather than a percentage value.

File: src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp
Comment (andrewmorell) [reply src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp]: I think it makes sense to reduce these long lines within 120 characters where they exist, applying that to each noted.

File: src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp
Comment (andrewmorell) [reply src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp]: Agreed, this has been updated.

File: src/simulation/dynamics/Thrusters/thrusterDynamicEffector/_UnitTest/test_thruster_integrated_sim.py
Comment (andrewmorell) [reply src/simulation/dynamics/Thrusters/thrusterDynamicEffector/_UnitTest/test_thruster_integrated_sim.py]: This has been done for this integrated test.

File: src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp
Comment (andrewmorell) [inline src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp:410]: One additional change has been made not already mentioned in these comments, based on a separate discussion between Cody and I. The thrust and Isp scaling factors for blow down have now been clamped to min = 0.0 and max = 1.0 as shown here.

File: src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp
Comment (codyallard28) [inline src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp:31]: Why is this -1.0 instead of 0.0?

**Review by codyallard28 — APPROVED** [review]
One small comment/question

File: src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp
Comment (andrewmorell) [reply src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp:31]: This ensures default behavior of blow down effects not being used if a fuel tank is never attached to update the fuelMass. I couldn't make 0.0 the default behavior because that is a real possibility for the fuel tank which would make the thruster think it is out of propellant and output zero thrust, so a negative value turns off blow down instead.

Comment (codyallard28) [reply src/simulation/dynamics/Thrusters/thrusterDynamicEffector/thrusterDynamicEffector.cpp:31]: After talking, I am good with this.

**Pull request merged**
